plugins/com.google.cloud.tools.eclipse.appengine.newproject.maven/src/com/google/cloud/tools/eclipse/appengine/newproject/maven/MavenAppEngineStandardWizardPage.java

```diff
@@ -323,4 +327,57 @@ public void modifyText(ModifyEvent event) {
       checkFlipToNext();
     }
   }
+
+  /**
+   * Auto-fills javaPackageField as "groupId.artifactId" (or "groupId" if artifactId is empty),
+   * only when 1) javaPackageField is empty; or 2) the field matches previous auto-fill before
+   * ID modification.
+   */
+  private final class AutoPackageNameSetterOnGroupIdChange implements VerifyListener {
+    @Override
+    public void verifyText(VerifyEvent event) {
+      // Below explains how to get text after modification:
+      // http://stackoverflow.com/questions/32872249/get-text-of-swt-text-component-before-modification
+      String newGroupId =
+          getGroupId().substring(0, event.start) + event.text + getGroupId().substring(event.end);
+
+      String oldPackageName = suggestPackageName(getGroupId(), getArtifactId());
+      String newPackageName = suggestPackageName(newGroupId, getArtifactId());
+      AutoSetPackageNameIfNecessary(oldPackageName, newPackageName);
+    }
+  }
+
+  /**
+   * See {@link AutoPackageNameSetterOnGroupIdChange}.
+   */
+  private final class AutoPackageNameSetterOnArtifactIdChange implements VerifyListener {
+    @Override
+    public void verifyText(VerifyEvent event) {
+      String newArtifactId = getArtifactId().substring(0, event.start)
+          + event.text + getArtifactId().substring(event.end);
+
+      String oldPackageName = suggestPackageName(getGroupId(), getArtifactId());
+      String newPackageName = suggestPackageName(getGroupId(), newArtifactId);
+      AutoSetPackageNameIfNecessary(oldPackageName, newPackageName);
+    }
+  }
+
+  /**
+   * See {@link AutoPackageNameSetterOnGroupIdChange#verifyText(VerifyEvent)}.
+   */
+  private void AutoSetPackageNameIfNecessary(String oldPackageName, String newPackageName) {
+    if (getPackageName().isEmpty() || getPackageName().equals(oldPackageName)) {
+      javaPackageField.setText(newPackageName);
+    }
+  }
+
+  /**
+   * Helper function returning a suggested package name based on groupId and artifactId.
+   */
+  private static String suggestPackageName(String groupId, String artifactId) {
+    if (!artifactId.isEmpty()) {
```

Comment:
Should filter out invalid characters (e.g., dashes) and lower-case everything.  Maven has very few rules for group and artifact IDs.